### ISSUE TITLE
How can you inject a service into an implementation of IClaimsTransformer?

### ISSUE THREAD

**Issue by csaloio** [aspnet/Security#718]
I need to be able to query a DB in my implementation.

My understanding is that you must add the transform like this in Startup.Configure:
```
app.UseClaimsTransformation(o => o.Transformer = new MyClaimsTransformer());
```

But how can it get access to a scoped service from within `MyClaimsTransformer` ?

**Comment by blowdart** [comment]
As of right now you can't. @HaoK we did talk about giving claims transformation more love ...

**Comment by csaloio** [comment]
@blowdart @HaoK I'm not sure how meaningful a transformation you can make without access to your data :(

**Comment by HaoK** [comment]
Since this is inside of Configure, you can service locate for it to work

```C#
var myTransformer = app.Services.GetRequiredService<MyTransformer>();
app.UseClaimsTransformation(o => o.Transformer = myTransformer);
```

**Comment by blowdart** [comment]
Kinda ugly though :)

**Comment by HaoK** [comment]
Also with the new options changes we can just make this work now that options will be a normal service.

**Comment by csaloio** [comment]
@HaoK I don't understand your example. `IApplicationBuilder` doesn't seem to have a `Services` property (at least in RC1). It does have 'ApplicationServices' but my understanding was that  his is only appropriate for services that have an application scope. I think it is important that any DB service injected into the transformer have request scope and not be scoped to the application.

**Comment by DovydasNavickas** [comment]
@HaoK, @blowdart, this really does not make sense, if the request is not exposed to the transformer. Any reasons why you cannot expose that?

**Comment by brockallen** [comment]
Seems like the `Options.Transformer` really should be pulled from DI.

**Comment by DovydasNavickas** [comment]
And within current request scope :clap:

**Comment by lasrol** [comment]
@csaloio getting db or usermanager from the ApplicationServices gave me a lot of open data reader errors

**Comment by csaloio** [comment]
@lasrol I don't suggest using ApplicationServices. What I think could be done (though it is **not ideal**), is to just new up your context inside of the `TransformAsync` method in the transformer making sure to wrap it in a using block so it is disposed of. Then use it read whatever data you need to transform the `ClaimsPrincipal`.

But as I said this is far from ideal. Everything about the .Net core design encourages the use of DI and but `IClaimsTransformer` appears to be a design oversight.

**Comment by Eilon** [comment]
This would indeed be nice but we are not taking many new features for this release.

**Comment by csaloio** [comment]
So will this be a breaking change in the future?

**Comment by DovydasNavickas** [comment]
@Eilon I understand why are you against taking new features into current release, but if you want to follow semver, this change would only go into >=2.0.0 release. And maybe the change needed here is quite simple: a reference to request, which by itself has `RequestServices`.

**Comment by DovydasNavickas** [comment]
And it wouldn't really break design of the feature, if I understand correctly. Because the transformation is either way done on every authenticated request.

**Comment by Eilon** [comment]
@HaoK / @Tratcher could/should we just add this property in the meantime?

**Comment by DovydasNavickas** [comment]
Oh, my bad. `Request` itself does not have `RequestServices`, they are on `HttpContext` and therefore `HttpContext` would need to be given.

And now that I look at the actual code and possible implementation, it might not be as easy as exposing `HttpContext` or `RequestServices`.

That could be done in the [ClaimsTransformationMiddleware](https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication/ClaimsTransformationMiddleware.cs#L44):
>```CSharp
public async Task Invoke(HttpContext context)
{
    var handler = new ClaimsTransformationHandler(Options.Transformer);
    handler.RegisterAuthenticationHandler(context.GetAuthentication());
    try
    {
        if (Options.Transformer != null)
        {
            context.User = await Options.Transformer.TransformAsync(context.User); // <-- Here
        }
        await _next(context);
    }
    finally
    {
        handler.UnregisterAuthenticationHandler(context.GetAuthentication());
    }
}
```

By giving `TransformAsync` method not only `context.User`, but also `context` itself or `context.RequestServices`.

But that would require changing `IClaimsTransformer` interface. And the interface is also used in [ClaimsTransformationHandler](https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication/ClaimsTransformationHandler.cs#L31) to pass on transformer and is called in `AuthenticateAsync`:
>```CSharp
public async Task AuthenticateAsync(AuthenticateContext context)
{
    if (PriorHandler != null)
    {
        await PriorHandler.AuthenticateAsync(context);
        if (_transform != null && context?.Principal != null)
        {
            context.Authenticated(
                await _transform.TransformAsync(context.Principal), // <-- Here
                context.Properties,
                context.Description);
        }
    }
}
```

In which case there is only `AuthenticateContext` available and no request, nor `HttpContext` is there.

Of course, as `ClaimsTransformationHandler` is only instantiated in [ClaimsTransformationMiddleware](https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNetCore.Authentication/ClaimsTransformationMiddleware.cs#L38), it is possible to pass `HttpContext` to its constructor:
```CSharp
public class ClaimsTransformationHandler : IAuthenticationHandler
{
    private readonly IClaimsTransformer _transform;
    private readonly HttpContext _httpContext;

    public ClaimsTransformationHandler(IClaimsTransformer transform, HttpContext httpContext)
    {
        _transform = transform;
        _httpContext = httpContext;
    }

    // <...>
}
```

And then pass the `HttpContext` to `TransformAsync`:
```CSharp
context.Authenticated(await _transform.TransformAsync(_httpContext, context.Principal), // <-- Here
                      context.Properties,
                      context.Description);
```

If you do not want to let the user have `HttpContext` itself, maybe the `HttpContext.RequestServices` is a good compromise? It would still allow to service locate required services, but in a request-scoped manner.

Or do you have something different in mind?

**Comment by csaloio** [comment]
@DovydasNavickas I believe that's a bit more of a service locator pattern than true DI, but it would still be vastly superior to what the current situation is.

**Comment by DovydasNavickas** [comment]
Yes, but as the default DI supports only constructor injection and currently `ClaimsTransformationMiddleware` takes `IClaimsTransformer` from `IOptions<ClaimsTransformationOptions>` in a constructor, I'm not sure request-scoped injection is possible here.

Which leaves service location that I described. It may not be the best possible solution here, but as you said - that would still be way better up until >= 2.0.0.

**Comment by csaloio** [comment]
I agree.

**Comment by Eilon** [comment]
Moving out of Backlog for now to make sure this gets discussed again.

**Comment by HaoK** [comment]
We can probably just make the IClaimsTransformer more like the IAuthEvent stuff, and hang the HttpContext off of the event context.

```
  public interface IClaimsTransformer {
     void Task<ClaimsPrincipal> TransformAsync(ClaimsTransformationContext context);
  }

  public class ClaimsTransformationContext {
    public HttpContext Context { get; }
    public ClaimsPrincipal Principal { get; }
  }
```

**Comment by csaloio** [comment]
That works too.

**Comment by DovydasNavickas** [comment]
Yep, that works :+1: I thought you wouldn't want to expose the whole HttpContext somewhy.
I can do PR for that.

**Comment by HaoK** [comment]
merged, thanks!  f5c1c45db20fd4592a3f49d4dc1d39dc5f62108c
